File: javascript/pilha.js
class PilhaCheiaErro extends Error {
    constructor(mensagem) {
        super(mensagem);
        this.name = "PilhaCheiaErro";
    }
}

class PilhaVaziaErro extends Error {
    constructor(mensagem) {
        super(mensagem);
        this.name = "PilhaVaziaErro";
    }
}

class TipoErro extends Error {
    constructor(mensagem) {
        super(mensagem);
        this.name = "TipoErro";
    }
}

class Pilha {
    constructor(capacidade, tipoDado) {
        this.capacidade = capacidade;
        this.tipoDado = tipoDado; // Ex: 'number', 'string'
        this.elementos = [];
    }

    empilha(dado) {
        if (this.pilha_esta_cheia()) {
            throw new PilhaCheiaErro("A pilha está cheia.");
        }
        if (typeof dado !== this.tipoDado) {
            throw new TipoErro(`O dado deve ser do tipo ${this.tipoDado}.`);
        }
        this.elementos.push(dado);
    }

    Desempilha() {
        if (this.pilha_esta_vazia()) {
            throw new PilhaVaziaErro("A pilha está vazia.");
        }
        return this.elementos.pop();
    }

    pilha_esta_vazia() {
        return this.elementos.length === 0;
    }

    pilha_esta_cheia() {
        return this.elementos.length >= this.capacidade;
    }

    troca() {
        if (this.tamanho() < 2) {
            throw new PilhaVaziaErro("Elementos insuficientes para a operação de troca.");
        }
        // Refatoração: Acesso direto aos índices do topo, eliminando pops/pushes redundantes
        const topoIdx = this.elementos.length - 1;
        const abaixoIdx = this.elementos.length - 2;

        const temp = this.elementos[topoIdx];
        this.elementos[topoIdx] = this.elementos[abaixoIdx];
        this.elementos[abaixoIdx] = temp;
    }

    tamanho() {
        return this.elementos.length;
    }
}

module.exports = { Pilha, PilhaCheiaErro, PilhaVaziaErro, TipoErro };

File: python/pilha.py

from array import array


class PilhaCheiaErro(Exception):
    pass


class PilhaVaziaErro(Exception):
    pass


class TipoErro(Exception):
    pass


class Pilha:
    def __init__(self, tipo, capacidade):
        # Validação do tipo
        if tipo not in (int, float, str):
            raise TipoErro(
                "A pilha só pode armazenar int, float ou str."
            )

        # Validação da capacidade
        if not isinstance(capacidade, int) or capacidade <= 0:
            raise ValueError(
                "A capacidade deve ser um inteiro positivo."
            )

        self.tipo = tipo
        self.capacidade = capacidade

       
        if tipo is int:
            self.dados = array('i')
        elif tipo is float:
            self.dados = array('d')
        else:
            self.dados = []


    def empilha(self, dado):
        # Primeiro verifica se o tipo do dado é válido
        if type(dado) is not self.tipo:
            raise TipoErro(
                "O dado possui tipo incompatível com a pilha."
            )

        # Depois verifica se há espaço
        if self.pilha_esta_cheia():
            raise PilhaCheiaErro(
                "A pilha está cheia."
            )

        self.dados.append(dado)


    def desempilha(self):
        if self.pilha_esta_vazia():
            raise PilhaVaziaErro(
                "A pilha está vazia."
            )

        return self.dados.pop()


    def pilha_esta_vazia(self):
        return not self.dados


    def pilha_esta_cheia(self):
        return len(self.dados) >= self.capacidade


    def troca(self):
        if self.tamanho() < 2:
            raise PilhaVaziaErro(
                "É necessário ter pelo menos dois elementos para trocar."
            )

        self.dados[-1], self.dados[-2] = (
            self.dados[-2],
            self.dados[-1]
        )


    def tamanho(self):
        return len(self.dados)
